### ISSUE TITLE
ImageName should be able to handle Image names when both tag and digest are provided

### ISSUE THREAD

**Issue by rohanKanojia** [eclipse/jkube#2545]
### Component

None

### Description

As described in this comment in [ORAS](https://github.com/oras-project/oras-go) [registry/reference.go](https://github.com/oras-project/oras-go/blob/3b1dd0e9700082fe7bc690a745263c06361417dd/registry/reference_test.go):

><--- path --------------------------------------------> |  - Decode `path`
<=== REPOSITORY ===> <--- reference ------------------> |    - Decode `reference`
<=== REPOSITORY ===> @ <=================== digest ===> |      - Valid Form A
<=== REPOSITORY ===> : <!!! TAG !!!> @ <=== digest ===> |      - Valid Form B (tag is dropped)
<=== REPOSITORY ===> : <=== TAG ======================> |      - Valid Form C
<=== REPOSITORY ======================================> |    - Valid Form D

When both tag and digest are provided, the tag should be ignored and digest should be used. Currently, ImageName is not able to handle this scenario and throws an exception.

### Expected Behavior
ImageName should be able to handle ImageName with both tag and digest (e.g.  `registry.example.com/hello-world:@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9`) provided and give precedence to digest.

### Acceptance Criteria
- [ ] ImageName is updated to handle scenarios including both tag and digest

**Comment by Devashishbasu** [comment]
Hi @rohanKanojia  i have a doubt, 
is IMAGE_NAME_REGEXP = Pattern.compile(NAME_COMPONENT_REGEXP + "(?:(?:/" + NAME_COMPONENT_REGEXP + ")+)?");
correct? because i can see an extra component "domainregexp" in its reference 

https://github.com/docker/docker/blob/04da4041757370fb6f85510c8977c5a18ddae380/vendor/github.com/docker/distribution/reference/regexp.go#L53

NameRegexp = expression(
		optional(DomainRegexp, literal(`/`)),
		nameComponentRegexp,
		optional(repeated(literal(`/`), nameComponentRegexp)))

**Comment by rohanKanojia** [comment]
@Devashishbasu : I think this regex only validates the name portion. Could you please check it by setting a breakpoint in `doValidate()` method and checking if `image` actually contains registry at during validation.  I think registry is already extracted from image name before reaching this path.

**Comment by Devashishbasu** [comment]
> @Devashishbasu : I think this regex only validates the name portion. Could you please check it by setting a breakpoint in `doValidate()` method and checking if `image` actually contains registry at during validation. I think registry is already extracted from image name before reaching this path.

yes, you are right registry is already extracted from image name before reaching this path.

but the IMAGE_NAME_REGEXP isn't allowing the dockers name without tag due to which this test without tag "registry.example.com/hello-world:@sha256:b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9" 
is failing 

java.lang.IllegalArgumentException: Given Docker name 'registry.example.com/hello-world:' is invalid:
   * image part 'hello-world:' doesn't match allowed pattern '[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?'


and for  tests which have both tag and digest,  according to this issue i am dropping the tag when digest is provided then also the test is failing   

ex-org.opentest4j.AssertionFailedError: 
expected: "quay.io/jkube/jkube-java:0.0.19@sha256:b7d8650e04b282b9d7b94daedf38321512f9910bce896cd40ffa15b1b92bab17"
 but was: "quay.io/jkube/jkube-java@sha256:b7d8650e04b282b9d7b94daedf38321512f9910bce896cd40ffa15b1b92bab17"


do i need to adjust the IMAGE_NAME_REGEXP to allow the dockers name without the tag?

**Comment by Devashishbasu** [comment]
> @Devashishbasu : I think this regex only validates the name portion. Could you please check it by setting a breakpoint in `doValidate()` method and checking if `image` actually contains registry at during validation. I think registry is already extracted from image name before reaching this path.

yes you are right , registry is already extracted from image name before reaching this path.
